dali/daliadmin/daliadmin.cpp

```diff
@@ -2794,114 +2798,177 @@ static void wuidCompress(const char *match, const char *type, bool compress)
+    }
 
-    PROGLOG("Gathering associated files");
-    conn.setown(querySDS().connect("/GeneratedDlls", myProcessSession(), fix?RTM_LOCK_WRITE:RTM_LOCK_READ, 10000));
-    if (!conn)
     {
-        PROGLOG("No generated DLLs associated with any workunit.\nExit. Took %d ms", ts.elapsed());
-        return;
+        PROGLOG("Gathering list of workunits");
+
+        ts.reset(0);
+        Owned<IRemoteConnection> workUnitConn = querySDS().connect("/WorkUnits", myProcessSession(), RTM_LOCK_READ, 10000);
+        if (!workUnitConn)
+            throw makeStringException(0, "Failed to connect to /WorkUnits");
+        Owned<IPropertyTreeIterator> wuidIter = workUnitConn->queryRoot()->getElements("*");
+        ForEach(*wuidIter)
+        {
+            IPropertyTree &wuid = wuidIter->query();
+            wuids.queryCreate(wuid.queryName());
+        }
+        PROGLOG("%u workunits gathered. Took %u ms", wuids.count(), ts.elapsed());
     }
-    IPropertyTree *root = conn->queryRoot()->queryBranch(NULL); // force all to download
+
+    std::vector<IPropertyTree *> generateDllsToDelete;
+
+    PROGLOG("Gathering associated files");
+
+    ts.reset(0);
+    Owned<IRemoteConnection> generatedDllsConn = querySDS().connect("/GeneratedDlls", myProcessSession(), fix?RTM_LOCK_WRITE:RTM_LOCK_READ, 10000);
+    if (!generatedDllsConn)
+        throw makeStringException(0, "Failed to connect to /GeneratedDlls");
+    IPropertyTree *root = generatedDllsConn->queryRoot()->queryBranch(NULL); // force everything to be fetched
 
     Owned<IPropertyTreeIterator> gdIter = root->getElements("*");
     RegExpr RE("^.*{W2[0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9]{-[0-9]+}?}{[^0-9].*|}$");
 
-    unsigned index=1;
+    unsigned removedLocations = 0;
+    unsigned totalCount = 0;
     ForEach(*gdIter)
     {
+        ++totalCount;
         IPropertyTree &gd = gdIter->query();
-        const char *name = gd.queryProp("@name");
-        if (name && *name)
+        const char *eName = gd.queryName();
+        const char *name = nullptr;
+        if (streq(eName, "GeneratedDll")) // legacy format
+        {
+            name = gd.queryProp("@name");
+            if (isEmptyString(name)) // would mean invalid entry if so
+                continue;
+        }
+        else if (startsWith(eName, "Dll")) // uniq entry name
+        {
+            name = gd.queryProp("@name");
+            if (isEmptyString(name)) // would mean invalid entry if so
+                continue;
+        }
+        else // unknown
+        {
+            continue;
+        }
+
+        bool associatedWithWuid = false;
+        bool entryRemoved = false;
+        if (RE.find(name)) // if it's a workunit associated dll/file
         {
-            if (RE.find(name))
+            StringBuffer wuid;
+            RE.substitute(wuid,"#1");
+            const char *w = wuid.str();
+            associatedWithWuid = (nullptr != wuids.find(*w));
```

Comment:
Is this code assuming that the wuid can be deduced from the dll name?  What happens if a workunit is cloned, and then the original is deleted?

Reply:
>Is this code assuming that the wuid can be deduced from the dll name?

it is.

>What happens if a workunit is cloned, and then the original is deleted?

good point. Does cloning like that still exist? A new workunit pointing to another workunits query so?

If so what it should do instead I think is walk all wuids and look at their associated files, to see if these are active dlls.
It can probably still assume that if there's a dll name to wuid name then active (as a shortcut), but if not a match, it can't assume not active..

Reply:
@jakesmith it occurs when you deploy a query and then execute that deployed query - the contents of the workunit are cloned (including the associated dll).  I have a feeling that I have seen it for Thor jobs that have been resubmitted as well.